crit/mempages.go

```diff
@@ -193,3 +194,55 @@ func (mr *MemoryReader) GetShmemSize() (int64, error) {
 
 	return size, nil
 }
+
+// SearchPattern searches for a pattern in the process memory pages.
+// If the pattern is found, it returns slices of strings containing the matched patterns
+// along with the specified context (number of bytes before and after each match).
+func (mr *MemoryReader) SearchPattern(pattern string, context int) ([]string, error) {
+	if context < 0 {
+		return nil, errors.New("context size cannot be negative")
+	}
+
+	regexPattern, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []string
+
+	for _, entry := range mr.pagemapEntries {
+		start := entry.GetVaddr()
+		end := start + uint64(entry.GetNrPages())*uint64(mr.pageSize)
+
+		// Read the entire pages content
+		buffer, err := mr.GetMemPages(start, end)
+		if err != nil {
+			return nil, err
+		}
+
+		// Replace null bytes with space
+		pagesContent := bytes.ReplaceAll(buffer.Bytes(), []byte{0x0}, []byte{0x20})
+
+		// Search for all matching pattern using the regular expression
+		matches := regexPattern.FindAllIndex(pagesContent, -1)
+
+		for _, match := range matches {
+			startPattern, endPattern := match[0], match[1]
+
+			// Calculate the start and end indices for the context surrounding the match
+			startContext := startPattern - context
+			if startContext < 0 {
+				startContext = startPattern
+			}
+
+			endContext := endPattern + context
+			if endContext > buffer.Len() {
+				endContext = endPattern
+			}
+
+			results = append(results, string(pagesContent[startContext:endContext]))
```

Comment:
One of the main benefits of using this functionality (in comparison to something like `grep`) is that it can display additional information about where a pattern is found in memory. Would it be possible to also include the virtual address in `results`?

Reply:
Sure. Would it make sense to create a type for each result match, for example:

```
// PatternMatch represents a match when searching for a regex pattern in memory.
type PatternMatch struct {
    Vaddr        uint64       // Virtual address of the first byte of the match
    Length       int           // Length of the pattern match in bytes
    Pattern      string       // String representation of the pattern match
}
```

Reply:
Sounds good